carriers/ups.js

```diff
@@ -0,0 +1,216 @@
+                if (address.city) {
+                    event.address.city = address.city;
+                }
+
+                if (address.state) {
+                    event.address.state = address.state;
+                }
+            }
+
+            results.events.push(event);
+        });
+
+        callback(null, results);
+    });
+}
+
+function UPS(options) {
+    this.isTrackingNumberValid = function(trackingNumber) {
+        // Remove whitespace
+        trackingNumber = trackingNumber.replace(/\s/g, '');
+
+        // https://www.ups.com/us/en/tracking/help/tracking/tnh.page
+        // https://www.codeproject.com/Articles/21224/Calculating-the-UPS-Tracking-Number-Check-Digit
+        if (/^1Z[0-9A-Z]{16}$/.test(trackingNumber)) {
+            if (confirmUps(trackingNumber)){
+                return true;
+            }
+        }
+        if (/^(H|T|J|K|F|W|M|Q|A)\d{10}$/.test(trackingNumber)) {
+            if (confirmUpsFreight(trackingNumber)){
+                return true;
+            }
+        }
+
+        return false;
+    };
+
+    this.track = function(trackingNumber, callback) {
+        const req = {
+            baseUrl: options.baseUrl || 'https://onlinetools.ups.com',
+            json: {
+                Security: {
+                    UPSServiceAccessToken: {
+                        AccessLicenseNumber: options.accessKey
+                    },
+                    UsernameToken: {
+                        Username: options.username,
+                        Password: options.password
+                    }
+                },
+                TrackRequest: {
+                    InquiryNumber: trackingNumber,
+                    Request: {
+                        RequestAction: 'Track',
+                        RequestOption: 'activity'
+                    }
+                }
+            },
+            method: 'POST',
+            timeout: 5000,
+            url: '/rest/Track'
+        };
+
+        async.retry(function(callback) {
+            request(req, callback);
+        }, function (err, res) {
+            if (err) {
+                return callback(err);
+            }
+
+            const results = {
+                events: []
+            };
+
+            if (!res.body.TrackResponse) {
+                if (res && res.body && res.body.Fault && res.body.Fault.detail && res.body.Fault.detail.Errors && res.body.Fault.detail.Errors.ErrorDetail && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode.Code === '151044') {
+                    // No Tracking Information
+                    return callback(null, results);
+                }
+
+                if (res && res.body && res.body.Fault && res.body.Fault.detail && res.body.Fault.detail.Errors && res.body.Fault.detail.Errors.ErrorDetail && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode && res.body.Fault.detail.Errors.ErrorDetail.PrimaryErrorCode.Description) {
```

Comment:
I can't think of one at the moment, but I wish there was a better way to do this that didn't involve so many expressions in the conditional. We do this sort of stuff often, but it's usually not so deep.
Not sure we need to change it, but I couldn't not say anything?